extensions/objects/templates/unicode-string-editor.component.spec.ts

```diff
@@ -0,0 +1,201 @@
+// Copyright 2021 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for unicode string editor.
+ */
+
+import { EventEmitter, SimpleChanges } from '@angular/core';
+import { ExternalSaveService } from 'services/external-save.service';
+import { FormsModule } from '@angular/forms';
+import { UnicodeStringEditorComponent } from './unicode-string-editor.component';
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+
+describe('UnicodeStringEditorComponent', () => {
+  let component: UnicodeStringEditorComponent;
+  let fixture: ComponentFixture<UnicodeStringEditorComponent>;
+  let externalSaveService: ExternalSaveService;
+
+  var externalSaveEventEmitter = new EventEmitter();
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      imports: [FormsModule],
+      declarations: [UnicodeStringEditorComponent]
+    }).compileComponents();
+  }));
+
+  beforeEach(angular.mock.module('oppia'));
+  beforeEach(() => {
+    fixture = TestBed.createComponent(UnicodeStringEditorComponent);
+    component = fixture.componentInstance;
+    externalSaveService = TestBed.get(ExternalSaveService);
+    spyOnProperty(externalSaveService, 'onExternalSave').and.returnValue(
+      externalSaveEventEmitter);
+    component.alwaysEditable = false;
+    component.value = 'random value';
+  });
+
+  afterEach(() => {
+    component.ngOnDestroy();
+  });
+
+  it('should initialize component when interaction rule is selected', () => {
+    spyOn(component, 'closeEditor').and.callThrough();
+    spyOn(component.componentSubscriptions, 'add');
+    component.active = true;
+
+    expect(component.active).toBe(true);
+
+    component.ngOnInit();
+
+    expect(component.componentSubscriptions.add).toHaveBeenCalled();
+    expect(externalSaveService.onExternalSave).toEqual(
+      externalSaveEventEmitter);
+    expect(component.componentSubscriptions.closed).toBe(false);
```

Comment:
All your post-checks should have corresponding pre-checks.

Reply:
I don't see the corresponding pre-check for e.g. alwaysEditable.

Please double-check to make sure that everything is covered.

Reply:
I had removed this line since it checks the same objective as line 63. The subscription is tested in a separate test.

Reply:
I don't follow -- https://github.com/oppia/oppia/pull/12996/files#diff-97532317345e77c56e41a813bb4f2ff9fce2f88ae76182f4d5ca91fab935de07R67 still has a missing pre-check.

Ah -- are you only doing it for this line? My comment was meant to apply to every line of every test, not just this one.

Reply:
Done.